### ISSUE TITLE
Question: Recommended row group size

### ISSUE THREAD

**Issue by vgalka-sl** [elastacloud/parquet-dotnet#392]
Hi guys,

I'm new to the parquet format. While going over the documentation, I found contradicting some recommendations.

On one hand, your ["Writing Data"](https://github.com/elastacloud/parquet-dotnet/blob/master/doc/writing.md) section states:
> It's not recommended to have more than 5'000 rows in a single row group for performance reasons.

On the other hand, the [official documentation on Apache site](https://parquet.apache.org/documentation/latest/) states:
> Row group size: Larger row groups allow for larger column chunks which makes it possible to do larger sequential IO. Larger groups also require more buffering in the write path (or a two pass write). We recommend large row groups (512MB - 1GB). Since an entire row group might need to be read, we want it to completely fit on one HDFS block. Therefore, HDFS block sizes should also be set to be larger. An optimized read setup would be: 1GB row groups, 1GB HDFS block size, 1 HDFS block per HDFS file.

When working with small rows, say 1KB, these two come in contradiction. 5000 rows x 1KB = 5MB, which is far from 0.5-1GB recommended by Apache. Would really appreciate some clarification on this.

Many thanks,
Vasili

**Comment by aloneguid** [comment]
Hi Vasili,

I think it rather depends on your data and underlying storage. HDFS used to be the most popular storage solution, however these days you would have files scattered in S3, azure blob storage etc. where block size varies.

5'000 rows is just a general average recommendation, you can't judge how much physical data it will take, as each data type has different storage requirements, compression capabilities etc. Multiple columns also contribute to data size.

It's up to you to experiment with your data and see how it fits best into underlying storage. This library doesn't force you with a particular partitioning strategy.

**Comment by xhochy** [comment]
Just to give some input on this one: We're using Parquet files sometimes with 5 million rows per RowGroup but also sometimes a bit smaller to have better predicate pushdown. They are stored on Azure Blob or S3. We have found that row group sizes below 30k make no difference anymore. The larger the row groups, the better compression works and this files are smaller. When you just read and write whole files: Only have a single RowGroup. When you only want to read parts of a file (predicate pushdown) or are in a memory-constrained situation, consider splitting your rows into more RowGroups.

Note: This is using `pyarrow.parquet` and not parquet-dotnet but similar figures apply probably.

**Comment by prdpsvs** [comment]
How to split rows into multiple row groups? I did not find a setting on parquet.

**Comment by jorgecarleitao** [comment]
@xhochy , when you mean a row group, are you assuming a single page per column chunk? Isn't the smallest compression unit a page, whose column chunks are composed of multiples?

**Comment by xhochy** [comment]
> when you mean a row group, are you assuming a single page per column chunk? 

Not at all, the statements were made independently of the number of pages per ColumnChunk. But at that time `pyarrow` also only supported RowGroup based predicate pushdown, so the number of pages was not a relevant size to tune for me.

> Isn't the smallest compression unit a page, whose column chunks are composed of multiples?

For "compression" in the Parquet meaning: yes. But you need to be aware that dictionary encoding is also doing a huge job in keeping the size down. This is done on a per-RowGroup and not per-page basis.

**Comment by jorgecarleitao** [comment]
> > Isn't the smallest compression unit a page, whose column chunks are composed of multiples?
> 
> For "compression" in the Parquet meaning: yes. But you need to be aware that dictionary encoding is also doing a huge job in keeping the size down. This is done on a per-RowGroup and not per-page basis.

I meant the statement 

> The larger the row groups, the better compression works and this files are smaller

Isn't this only valid if we either use dict-encoding or a single page per row group?

**Comment by xhochy** [comment]
> I meant the statement
> 
> > The larger the row groups, the better compression works and this files are smaller
> 
> Isn't this only valid if we either use dict-encoding or a single page per row group?

These two greatly improve it but also increasing the size of the row group limits the fragmentation of pages if you have more than a page per RowGroup. I have seen users using row groups of 1024 rows. There, multiple row groups could be combined into a single page if row groups were larger. These effects are though smaller as the two other mentioned.
